### ISSUE TITLE
IP behind multiple Proxies is being returned as 0.0.0.0

### ISSUE THREAD

**Issue by Ipstenu** [johngodley/redirection#2847]
This is likely related to https://github.com/johngodley/redirection/issues/210

### Summary

Following an update to our proxy stack, this plugin returns IP 0.0.0.0 for all transactions.

Looking at our logs, we're properly sending headers:

```
X-Forwarded-For: VALIDIP, 0.0.0.0
Cf-Connecting-Ip: VALIDIP
```

We would expect that to return the VALIDIP, but instead it's 0's.

Looking at the code:

```
                foreach ( self::get_ip_headers() as $var ) {
                        if ( ! empty( $_SERVER[ $var ] ) ) {
                                $ip = $_SERVER[ $var ];
                                $ip = explode( ',', $ip );
                                $ip = array_shift( $ip );
                                break;
                        }
                }
```

That foreach loop is checking everything from get_ip_headers() and in theory it should be breaking once it grabs HTTP_CF_CONNECTING_IP, but it's not.

I would recommend changing this: `if ( ! empty( $_SERVER[ $var ] ) )`

to this: `if ( ! empty( $_SERVER[ $var ] ) && '' !== $ip )`

Basically I think the issue is that the BREAK is in the IP check, but the for-loop keeps running so it overwrites. I know it's supposed to break the for-loop, but it appears that it's not doing it consistently.

**Comment by johngodley** [comment]
I've not been able to recreate this problem, and the extra check shouldn't make any difference. The only way I've been able to make this fail is if `VALIDIP` is itself `0.0.0.0`

Does the change fix it for you?

If you go to the 'support' page in the plugin, what does it show for the IP Headers? It should look something like this:

![image](https://user-images.githubusercontent.com/1277682/113470957-1aee0380-9451-11eb-8fdc-39c67e0e6c16.png)

**Comment by terryheff** [comment]
Hi John

This issue was raised on my behalf by my hosting provider. In answer to your question. On the options page, IP Headers:
![image](https://user-images.githubusercontent.com/42607181/113502048-8ef7dc80-9521-11eb-9120-389950940f45.png)
 Hope this helps

Regards

**Comment by johngodley** [comment]
Based on that it seems like your `HTTP_CF_CONNECTING_IP` value is `0.0.0.0`. This value isn't processed in anyway, and isn't part of the above mentioned loop.

**Comment by terryheff** [comment]
Ok - but that leaves me wondering what the point of your question is. As a user and as far as I can see, I have no control over the IP Headers field. 

Something has obviously changed to cause this. If it's any help, the issue first appeared on 26th March. Prior to that it was working fine. How can this issue be rectified?

**Comment by johngodley** [comment]
I asked so I could see if the problem is caused by the plugin. Based on the above it seems like Redirection is behaving as expected - you have `0.0.0.0` and so the plugin is recording `0.0.0.0`.

You will need to configure your server so it correctly puts the IP address in `HTTP_CF_CONNECTING_IP`. Failing that, you can delete the variable entirely and it will then use `REMOTE_ADDR`.

If you cannot change the server then you can do it in PHP in your `wp-config.php` file.

**Comment by Ipstenu** [comment]
Actually it's putting in MULTIPLE IPs, one of which is in fact 0.0.0.0

Like I said: X-Forwarded-For: VALIDIP, 0.0.0.0

That first item? It's the VALIDIP. 1.2.3.4 whatever. It's being ignored in favour of 0.0.0.0

So the issue is your plugin cheerfully ignores the valid IP that came first and picks up the second one.

**Comment by johngodley** [comment]
Based on the above details, `HTTP_X_FORWARDED_FOR` is not present, but `REMOTE_ADDR` is. Regardless, `HTTP_CF_CONNECTING_IP` is taking priority. As `HTTP_CF_CONNECTING_IP` is `0.0.0.0` then that's the value that is used.

If you cannot modify your server to insert the correct IP into `HTTP_X_FORWARDED_FOR` then you can change this in PHP by modifying your `wp-config.php` and setting/deleting the header value directly. Additionally you can use the WP hook `redirection_request_ip` to override whatever IP is used.

`$_SERVER['HTTP_CF_CONNECTING_IP'] = $_SERVER['REMOTE_ADDR'];`

Or:

`unset( $_SERVER['HTTP_CF_CONNECTING_IP'] );`

**Comment by terryheff** [comment]
Hello John

I've used unset() in wp-config.php and so far it seems to be working. Thanks for you help

Regards

**Comment by Ipstenu** [comment]
> Based on the above details, HTTP_X_FORWARDED_FOR is not present, but REMOTE_ADDR is. Regardless, HTTP_CF_CONNECTING_IP is taking priority. As HTTP_CF_CONNECTING_IP is 0.0.0.0 then that's the value that is used.

No, again: **X-Forwarded-For: VALIDIP, 0.0.0.0**

X-Forwarded-For has two IPs, a valid one and 0.0.0.0

Your plugin is (for whatever reason) giving priority to the second IP in HTTP_X_FORWARDED_FOR, however, as (again) I said in the original post: **Cf-Connecting-Ip: VALIDIP**

So it's _NOT_ giving priority to HTTP_CF_CONNECTING_IP, HTTP_CF_CONNECTING_IP is properly set to a valid IP.

**Comment by johngodley** [comment]
I'm a little confused. Are both the issues described above for the same site, or different sites? I was under the impression you were working together.

If you are not then please can you also provide the IP headers output as originally requested? I am unable to reproduce any problem with the processing of IP addresses based on your description. Even when you write it again.

**Comment by Ipstenu** [comment]
The same site, in that instance, but we do have a couple other people. Making the $ip change I suggested resolved it for the one person who allowed me to make that edit. They aren't my sites :) I have to ask.

> If you go to the 'support' page in the plugin, what does it show for the IP Headers? It should look something like this:

Assuming you mean that, I have to ask permission to log in (and I m waiting on someone else for that).

That said, our code **does not** prioritize HTTP_CF_CONNECTING_IP

Look here:

https://github.com/johngodley/redirection/blob/a2af44bfc69a7ab7656cbe3b8f706cad3bb742d3/models/request.php#L105

See how the _first_ item is HTTP_CF_CONNECTING_IP ?

Now in theory, having the break here - https://github.com/johngodley/redirection/blob/a2af44bfc69a7ab7656cbe3b8f706cad3bb742d3/models/request.php#L127 - will cause it to stop processing. But for whatever reason it's still doing. So if you put in the `'' !== $ip ` into the loop on https://github.com/johngodley/redirection/blob/a2af44bfc69a7ab7656cbe3b8f706cad3bb742d3/models/request.php#L122 then it will be a fallback for cases where, for whatever reason, it fails to break.

Naturally you won't see a change if that's not your issue.

I'm happy to tell the customers to put in that override, but in the error logs and the access logs and everything I have permission to look at, HTTP_CF_CONNECTING_IP is absolutely 100% sending the IP as it should.

**Comment by johngodley** [comment]
I've run the code through every possible combination of IPs I can think of and it behaves as expected. I have stepped through the code with a debugger and it also works as expected and the `break` always exits the loop. Even if the `break` is removed it still wouldn't pick `0.0.0.0` and would use the first address in `HTTP_X_FORWARDED_FOR`.

The first example on the [PHP page](https://www.php.net/manual/en/control-structures.break.php) for `break` shows it being used in the same way.

At the moment I have no reason to think that PHP is behaving in any way different to what is expected. It seems more likely that, as in the [above instance](https://github.com/johngodley/redirection/issues/2847#issuecomment-813376162), the IP address is incorrectly set.

Your access and error logs may contain the correct IP address, but that value is logged directly by the server, not PHP, and does not indicate that PHP has the correct IP.

If you have a reproducible example that demonstrates the problem then I will happily make a change - I'm really not trying to be difficult here. It is something I can then add to one of the unit tests and make sure it doesn't happen again.

If you can provide a screenshot of the IP Header information then I can use the exact same values.

**Comment by johngodley** [comment]
I'll close this for now, please reopen with more information
